msresist/clustering.py

```diff
@@ -129,18 +131,40 @@ def pssms(self, PsP_background=False):
         return pssms
 
     def predict_UpstreamKinases(self):
-        """Compute matrix-matrix similarity between kinase specificity profiles and cluster PSSMs to identify upstream kinases regulating clusters."""
-        PSPLs = PSPSLdict()
-        PSSMs = [np.delete(np.array(list(np.array(mat))), [5, 10], 1) for mat in self.pssms(PsP_background=True)]  # Remove P0 and P+5 from pssms
-        a = np.zeros((len(PSPLs), len(PSSMs)))
-
-        for ii, spec_profile in enumerate(PSPLs.values()):
-            for jj, pssm in enumerate(PSSMs):
-                a[ii, jj] = np.linalg.norm(pssm - spec_profile)
-
-        table = pd.DataFrame(a)
-        table.insert(0, "Kinase", list(PSPSLdict().keys()))
-        return table
+        """Use multi-dimensional scaling to match kinase profiling with cluster motifs."""
+        pspls = list(PSPSLdict().values())
+        pssms = [np.delete(np.array(list(np.array(mat))), [5, 10], 1) for mat in self.pssms(PsP_background=True)]
+        mats = pspls + pssms
+
+        n = len(mats)
+        res = np.empty((n, n), dtype=float)
+        for ii in range(n):
+            for jj in range(n):
+                res[ii, jj] = np.linalg.norm(mats[ii] - mats[jj])
+
+        res[res < 1.0e-100] = 0
+
+        seed = np.random.RandomState(seed=3)
+        mds = MDS(n_components=2, max_iter=3000, eps=1e-9, random_state=seed,
+                        dissimilarity="precomputed", n_jobs=1)
+        pos = mds.fit(res).embedding_
+
+        nmds = MDS(n_components=2, metric=False, max_iter=3000, eps=1e-12,
+                            dissimilarity="precomputed", random_state=seed, n_jobs=1,
+                            n_init=1)
+        npos = nmds.fit_transform(res, init=pos)
+
+        clf = PCA(n_components=2)
+        npos = clf.fit_transform(npos)
+        varExp = np.round(clf.explained_variance_ratio_, 2)
```

Comment:
The PCA is of the already reduced MDS space, so you can't provide a variance explained.

Also, you might want to look at components 3 and 4. Those might separate the remaining kinases bunched together.

Reply:
Ok, variance explained removed.

This is how it looks like fitting MDS, NMDS, and PCA with 4 components:

![Screen Shot 2020-11-24 at 2 08 01 PM](https://user-images.githubusercontent.com/46428799/100156838-7a625400-2e5e-11eb-847e-04f61975ba8d.png)